services/galley/src/Galley/API/Federation.hs

```diff
@@ -529,26 +533,58 @@ instance
 
 mlsSendWelcome ::
   Members
-    '[ GundeckAccess,
+    '[ BrigAccess,
+       GundeckAccess,
        Input (Local ()),
        Input UTCTime
      ]
     r =>
   Domain ->
   F.MLSWelcomeRequest ->
-  Sem r ()
-mlsSendWelcome _origDomain (F.MLSWelcomeRequest b64RawWelcome rcpts) = do
+  Sem r F.MLSWelcomeResponse
+mlsSendWelcome _origDomain (F.unMLSWelcomeRequest -> welBase64) = do
   loc <- input @(Local ())
   now <- input @UTCTime
-  let rawWelcome = fromBase64ByteString b64RawWelcome
-  void $
-    runMessagePush loc Nothing $
-      foldMap (uncurry $ mkPush rawWelcome loc now) (F.unMLSWelRecipient <$> rcpts)
+  let rawWelcome = fromBase64ByteString welBase64
+  runErrors $ do
+    welcome <-
+      fromEither
+        . mapLeft (const . mlsProtocolError $ decodingErrorMsg)
+        . decodeMLS'
+        $ rawWelcome
+    -- Extract only recipients local to this backend
+    rcpts <-
+      fmap fold $
+        foldQualified loc ((: []) . tUnqualified) (const mempty)
+          <$$> welcomeRecipients welcome
+    void $
+      runMessagePush loc Nothing $
+        foldMap (uncurry $ mkPush rawWelcome loc now) rcpts
   where
+    decodingErrorMsg = "Could not decode the welcome message"
     mkPush :: ByteString -> Local x -> UTCTime -> UserId -> ClientId -> MessagePush 'Broadcast
     mkPush rawWelcome l time u c =
       -- FUTUREWORK: use the conversation ID stored in the key package mapping table
       let lcnv = qualifyAs l (Data.selfConv u)
           lusr = qualifyAs l u
           e = Event (qUntagged lcnv) (qUntagged lusr) time $ EdMLSWelcome rawWelcome
        in newMessagePush l () Nothing defMessageMetadata (u, c) e
+
+    runErrors ::
+      Sem (Error MLSProtocolError ': ErrorS 'MLSKeyPackageRefNotFound ': r) () ->
+      Sem r F.MLSWelcomeResponse
+    runErrors = fmap F.MLSWelcomeResponse . runDerefError . runProtocolError
+
+    runProtocolError ::
+      Sem (Error MLSProtocolError ': r) () ->
+      Sem r (Either F.MLSWelcomeResponseError ())
+    runProtocolError =
+      fmap (mapLeft (\(Tagged msg) -> F.MLSWelcomeResponseErrorDecodingFailed msg))
+        . runError
+
+    runDerefError ::
+      Sem (ErrorS 'MLSKeyPackageRefNotFound ': r) (Either F.MLSWelcomeResponseError ()) ->
+      Sem r (Either F.MLSWelcomeResponseError ())
+    runDerefError =
+      fmap (join . mapLeft (const F.MLSWelcomeResponseErrorRefNotFound))
+        . runError
```

Comment:
Could you condense this error handling into just one composition chain? See for example https://github.com/wireapp/wire-server/pull/2378/files#diff-124e9b9cb198ffe2963c75b0fd6233a4c89d690eaa0eb4c76dc80bf75fa17d8eR546-R551.

Reply:
This was taken care of yesterday as well.